subprojects/core/src/main/java/org/gradle/deployment/internal/DefaultDeploymentRegistry.java

```diff
@@ -17,61 +17,128 @@
         try {
             failIfStopped();
-            if (!handles.containsKey(id)) {
-                handles.put(id, handle);
+            if (!handles.containsKey(name)) {
+                return buildOperationExecutor.call(new CallableBuildOperation<T>() {
+                    @Override
+                    public BuildOperationDescriptor.Builder description() {
+                        return BuildOperationDescriptor.displayName("Start deployment '" + name + "'");
+                    }
+
+                    @Override
+                    public T call(BuildOperationContext context) {
+                        T delegate = objectFactory.newInstance(handleType, params);
+                        DeploymentHandleWrapper handle = new DeploymentHandleWrapper(name, delegate);
+                        if (pendingChanges.hasRemainingChanges()) {
+                            handle.pendingChanges(true);
+                        }
+                        handles.put(name, handle);
+                        return delegate;
+                    }
+                });
             } else {
-                throw new IllegalStateException("A deployment with id '" + id + "' is already registered.");
+                throw new IllegalStateException("A deployment with id '" + name + "' is already registered.");
             }
         } finally {
             lock.unlock();
         }
     }
 
     @Override
-    public <T extends DeploymentHandle> T get(Class<T> handleType, String id) {
+    public <T extends DeploymentHandle> T get(String name, Class<T> handleType) {
         lock.lock();
         try {
             failIfStopped();
-            return Cast.cast(handleType, handles.get(id));
+            if (handles.containsKey(name)) {
+                return Cast.cast(handleType, handles.get(name).getDelegate());
+            } else {
+                return null;
+            }
         } finally {
             lock.unlock();
         }
     }
 
     @Override
-    public void onNewBuild(Gradle gradle) {
+    public void onPendingChanges() {
+        lock.lock();
+        try {
+            pendingChanges.changesMade();
+            for (DeploymentHandle handle : handles.values()) {
+                handle.pendingChanges(true);
+            }
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    public void buildFinished(BuildResult buildResult) {
         lock.lock();
         try {
+            pendingChanges.changesIncorporated();
             for (DeploymentHandle handle : handles.values()) {
-                handle.onNewBuild(gradle);
+                Throwable failure = buildResult.getFailure();
+                if (failure == null) {
+                    handle.buildSucceeded();
+                } else {
+                    handle.buildFailed(failure);
+                }
+                if (!pendingChanges.hasRemainingChanges()) {
+                    handle.pendingChanges(false);
+                }
```

Comment:
This section feels wrong to me. Aren't we sending a 'reload now' signal with `buildSucceeded()` and then immediately sending a 'hang on, don't reload, we have pending changes' signal?

I can see 2 options that could be simpler:
- Don't even tell the deployment about a build completed when there are still pending changes. Just wait until we can send the "new deployment available" signal, which means that it's ready to reload.
- Tell the deployment about both the 'build compete/failure' and the 'there are pending changes' in the same call. Splitting these between calls will make the receiver more complex, I think.